resourcemanager/gpool/spool/spool.go

```diff
@@ -0,0 +1,163 @@
+
+	"github.com/pingcap/tidb/metrics"
+	"github.com/pingcap/tidb/resourcemanager"
+	"github.com/pingcap/tidb/resourcemanager/pooltask"
+	"github.com/pingcap/tidb/resourcemanager/util"
+	"github.com/pingcap/tidb/util/gpool"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.uber.org/zap"
+)
+
+// Pool is a single goroutine pool. it can not reuse the goroutine.
+type Pool struct {
+	wg                 sync.WaitGroup
+	mu                 sync.Mutex
+	options            *Options
+	capacity           atomic.Int32
+	running            atomic.Int32
+	isStop             atomic.Bool
+	concurrencyMetrics prometheus.Gauge
+	taskManager        pooltask.TaskManager[any, any, any, any, pooltask.NilContext]
+	gpool.BasePool
+}
+
+// NewPool create a single goroutine pool.
+func NewPool(name string, size int32, component util.Component, options ...Option) (*Pool, error) {
+	opts := loadOptions(options...)
+	result := &Pool{
+		options:            opts,
+		concurrencyMetrics: metrics.PoolConcurrencyCounter.WithLabelValues(name),
+		taskManager:        pooltask.NewTaskManager[any, any, any, any, pooltask.NilContext](size), // TODO: this general type
+	}
+	result.SetName(name)
+	result.capacity.Store(size)
+	result.concurrencyMetrics.Set(float64(size))
+	err := resourcemanager.InstanceResourceManager.Register(result, name, component)
+	if err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+// Tune changes the pool size.
+func (p *Pool) Tune(size int) {
+	p.SetLastTuneTs(time.Now())
+	p.capacity.Store(int32(size))
+	p.concurrencyMetrics.Set(float64(size))
+}
+
+// Run runs a function in the pool.
+func (p *Pool) Run(fn func()) error {
+	if p.isStop.Load() {
+		return gpool.ErrPoolClosed
+	}
+	_, run := p.isRunnable(1, true)
+	if !run {
+		return gpool.ErrPoolOverload
+	}
+	p.taskManager.RegisterTask(p.NewTaskID(), 1)
+	p.run(fn)
+	return nil
+}
+
+// Cap returns the capacity of the pool.
+func (p *Pool) Cap() int {
+	return int(p.capacity.Load())
+}
+
+// Running returns the number of running goroutines.
+func (p *Pool) Running() int {
+	return int(p.running.Load())
+}
+
+func (p *Pool) run(fn func()) {
+	p.wg.Add(1)
+	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				logutil.BgLogger().Error("recover panic", zap.Any("recover", r))
+			}
+			p.wg.Done()
+			p.running.Add(-1)
+		}()
+		fn()
+	}()
+}
+
+// RunWithConcurrency runs a function in the pool with concurrency.
+func (p *Pool) RunWithConcurrency(fn func(), concurrency int) error {
+	if p.isStop.Load() {
+		return gpool.ErrPoolClosed
+	}
+	conc, run := p.isRunnable(int32(concurrency), false)
+	if !run {
+		return gpool.ErrPoolOverload
+	}
+	// TODO: taskManager need to refactor
+	p.taskManager.RegisterTask(p.NewTaskID(), conc)
+	for n := int32(0); n < conc; n++ {
+		p.run(fn)
+	}
+	return nil
+}
+
+// isRunnable is to check if the pool is runnable.
+func (p *Pool) isRunnable(concurrency int32, mustEnough bool) (conc int32, run bool) {
```

Comment:
Will it wait until the pool is runnable (without `p.optons.Nonblocknig`)? How about renaming it to `waitxxxx`, or adding a new function to implement `wait`.

Reply:
> `Nonblocknig`

Yes, It will wait. so It rename it to ```check``` and ```checkInternal```.